### PR TITLE
[ONNX] Preprocess index_put with bool inputs to masked_scatter/masked_fill

### LINKED ISSUE #34054: Export to ONNX models including logical masks
## 🐛 Bug

This is a bug in torch.onnx module during export. Unable to handle with a logical mask due to TypeError.

## To Reproduce

1. type to the forward method something like
> mask = input_mask.clone()
> mask[mask != some_constant] = 1
> mask[mask == some_constant] = 0
2. export the model to onnx: 
>torch.onnx.export(model, dummy_input, filename, verbose=True, opset_version=11)

3. load the model by 
>onnxruntime.InferenceSession(filename)

4. catch the error. 
>This is an invalid model. Type Error: Type 'tensor(bool)' of input parameter (337) of operator (ScatterND) in node () is invalid.

5. see the log
>...
  %334 : Bool(20, 50, 6) = onnx::Cast[to=9](%333) # C:\Users\....\file.py:49:0
  %335 : Long() = onnx::Constant[value={1}]() # C:\Users\....\file.py:49:0
  %336 : Tensor = onnx::Shape(%334)
  %337 : Tensor = onnx::Unsqueeze[axes=[-1]](%334)
  %338 : Tensor = onnx::Shape(%atom_degree_list)
  %339 : Tensor = onnx::Constant[value={0}]()
  %340 : Tensor = onnx::Constant[value={1}]()
  %341 : Tensor = onnx::Constant[value={9223372036854775807}]()
  %342 : Tensor = onnx::Slice(%338, %340, %341, %339)
  %343 : Tensor = onnx::Concat[axis=0](%336, %342)
  %344 : Tensor = onnx::Reshape(%335, %343)
  %345 : Long(20, 50, 6) = onnx::ScatterND(%input_mask, %337, %344) # C:\Users\....\file.py:49:0
...

## Expected behavior
The workaround:
I can alter the code in the forward method like 
> mask = input_mask.clone()
> mask = mask - some_constant
> mask = numpy.heaviside(mask.abs(), 0).long()

and load the exported model successfully by onnxruntime. 
But I'm trying to export a network from a third-party library. Therefore I'm not allowed to alter the code.

## Environment

 - PyTorch Version : 1.4.0
 - OS: Windows 10 Pro 64 bit
 - How you installed PyTorch (`conda`, `pip`, source): current repo (git clone ...)
 - Build command you used (if compiling from source): setup.py build/install
 - Python version: 3.6
 - CUDA/cuDNN version: None
 - GPU models and configuration: None
 - Any other relevant information: opset 11

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

### PR DESCRIPTION
**Pull request by shubhambhokare1**
When the input to an indexing operation is a boolean, for example array[True] = value,
the subsequent index_put node formed needs to be converted to masked_scatter/masked_fill node based on the type of val the indexing node is equated. If that value is just a single scalar, then we use the masked_fill functionality and if value is a tensor of appropriate size, we use the masked_scatter functionality.

Fixes #34054

### CODE REVIEW

**Comment by KsenijaS** [comment]
Can you add a better description, also try to explain why and when you use aten::masked_scatter and aten::masked_fill

File: torch/csrc/jit/passes/onnx/preprocess_for_onnx.cpp [hunk 1]
```diff
@@ -158,11 +158,89 @@ static void ReplaceAddWithConcat(Block* b) {
...
+      ReplaceIndexPutWithMaskedScatter(child_block);
+    }
+    if (it->kind() == aten::index_put_) {
+      auto* lc_node = it->input(1)->node();
+      if (!(lc_node->input(0)->type()->cast<TensorType>()) ||
+          (lc_node->input(0)
+               ->type()
+               ->cast<TensorType>()
+               ->scalarType()
+               .value()) != c10::ScalarType::Bool) {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/preprocess_for_onnx.cpp:207]: This `.value()` needs to be safe guard with `.has_value()`, or checking that `...->scalarType()` is not `c10::nullopt`

File: torch/csrc/jit/passes/onnx/preprocess_for_onnx.cpp [hunk 2]
```diff
@@ -158,11 +158,89 @@ static void ReplaceAddWithConcat(Block* b) {
...
+      if ((!lc_node->inputs().size()) == 1) {
+        continue;
+      }
+
+      // If equated value is just a single scalar, then convert to masked_fill,
+      // and if value is a tensor of appropriate size, we convert to
+      // masked_scatter.
+      Node* masked_node;
+      if (it->input(2)->type()->cast<TensorType>() &&
+          (it->input(2)->type()->cast<TensorType>()->sizes().size().value()) ==
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/preprocess_for_onnx.cpp:220]: Same as above for `.value()`

**Comment by Enolerobotti** [comment]
Thank you @shubhambhokare1! This patch actually resolves #34054. @apaszke I would recommend this patch

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!

**Review by bzinodev — APPROVED** [review]

**Comment by shubhambhokare1** [comment]
@bzinodev Any changes required on my end for the PR to be merged?

**Pull request closed without merging**